File: src/solutions/2024/03/csharp/Main.cs

using System.Reflection.Metadata;
using System.Text.RegularExpressions;
using Core;

namespace Year2024;

public static class Day03
{
   static SolutionOptions OPTIONS = new()
   {
      HasAlternate = true,
      HasIO = false,
   };

   static string Solve(string input, bool p2)
   {
      var instructed = true;
      return new Regex("(mul\\(\\d{1,3},\\d{1,3}\\)|do\\(\\)|don't\\(\\))")
         .Matches(input)
         .Select(m =>
         {
            if (m.Value == "do()") { if (p2) instructed = true; return 0; }
            if (m.Value == "don't()") { if (p2) instructed = false; return 0; }
            if (!instructed) { return 0; }
            var val = new Regex("\\d{1,3}").Matches(m.Value);
            return int.Parse(val[0].Value) * int.Parse(val[1].Value);
         })
         .Sum()
         .ToString();
   }

   static string Part1(string input, bool isTest = false)
   {
      return Solve(input, false);
   }

   static string Part2(string input, bool isTest = false)
   {
      return Solve(input, true);
   }

   static void Main(string[] args)
   {
      Runner.Run(args, Part1, Part2, OPTIONS);
   }
}

File: src/solutions/2024/03/ts/main.ts
import { argv } from "node:process";
import type { SolutionOptions } from "src/options.ts";
import { run } from "src/run.ts";

const options: SolutionOptions = {
   hasAlternate: true,
   hasIo: false,
};

function solve(input: string, p2: boolean): string {
   let doit = true;
   let mul = 0;
   for (const m of input.match(/(mul\(\d{1,3},\d{1,3}\)|do\(\)|don't\(\))/g)!) {
      if (m === "do()") {
         if (p2) doit = true;
         continue;
      }
      if (m === "don't()") {
         if (p2) doit = false;
         continue;
      }
      if (!doit) continue;
      const [_, a, b] = m.match(/(\d{1,3}),(\d{1,3})/)!;
      mul += parseInt(a) * parseInt(b);
   }
   return mul.toString();
}

function part1(input: string, _isTest: boolean): string {
   return solve(input, false);
}

function part2(input: string, _isTest: boolean): string {
   return solve(input, true);
}

if (import.meta.main) {
   run(argv, part1, part2, options);
}
